src/lua/datetime.lua

```diff
@@ -0,0 +1,623 @@
+    local len = #range
+    assert(len == 2 or len == 3)
+
+    local left, right, neg = unpack(range)
+    if neg == v or (v >= left and v <= right) then
+        return
+    end
+
+    if neg == nil then
+        error(('value %d of %s is out of allowed range [%d, %d]'):
+              format(v, txt, left, right), 2)
+    else
+        error(('value %d of %s is out of allowed range [%d, %d..%d]'):
+              format(v, txt, neg, left, right), 2)
+    end
+end
+
+local function nyi(msg)
+    local text = 'Not yet implemented'
+    if msg ~= nil then
+        text = ("%s : '%s'"):format(text, msg)
+    end
+    error(text, 3)
+end
+
+-- offset if __seconds__ of 1970-01-01 from 0000-01-01
+local SECS_EPOCH_OFFSET = (DT_EPOCH_1970_OFFSET * SECS_PER_DAY)
+
+-- convert from epoch related time to Rata Die related
+local function local_rd(secs)
+    return math_floor((secs + SECS_EPOCH_OFFSET) / SECS_PER_DAY)
+end
+
+-- convert UTC seconds to local seconds, adjusting by timezone
+local function local_secs(obj)
+    return obj.epoch + obj.tzoffset * 60
+end
+
+local function utc_secs(epoch, tzoffset)
+    return epoch - tzoffset * 60
+end
+
+-- get epoch seconds, shift to the local timezone
+-- adjust from 1970-related to 0000-related time
+-- then return dt in those coordinates (number of days
+-- since Rata Die date)
+local function local_dt(obj)
+    return builtin.tnt_dt_from_rdn(local_rd(local_secs(obj)))
+end
+
+local function normalize_nsec(secs, nsec)
+    if nsec < 0 then
+        secs = secs - 1
+        nsec = nsec + NANOS_PER_SEC
+    elseif nsec >= NANOS_PER_SEC then
+        secs = secs + 1
+        nsec = nsec - NANOS_PER_SEC
+    end
+    return secs, nsec
+end
+
+local function datetime_cmp(lhs, rhs)
+    if not is_datetime(lhs) or not is_datetime(rhs) then
+        return nil
+    end
+    local sdiff = lhs.epoch - rhs.epoch
+    return sdiff ~= 0 and sdiff or (lhs.nsec - rhs.nsec)
+end
+
+local function datetime_eq(lhs, rhs)
+    local rc = datetime_cmp(lhs, rhs)
+    return rc ~= nil and rc == 0
+end
+
+local function datetime_lt(lhs, rhs)
+    local rc = datetime_cmp(lhs, rhs)
+    return rc == nil and error('incompatible types for comparison', 2) or
+           rc < 0
+end
+
+local function datetime_le(lhs, rhs)
+    local rc = datetime_cmp(lhs, rhs)
+    return rc == nil and error('incompatible types for comparison', 2) or
+           rc <= 0
+end
+
+local function datetime_serialize(self)
+    return { epoch = self.epoch, nsec = self.nsec,
+             tzoffset = self.tzoffset, tzindex = 0 }
+end
+
+local parse_zone
+
+local function datetime_new_raw(epoch, nsec, tzoffset, tzindex)
+    local dt_obj = ffi.new(datetime_t)
+    dt_obj.epoch = epoch
+    dt_obj.nsec = nsec
+    dt_obj.tzoffset = tzoffset or 0
+    dt_obj.tzindex = tzindex or 0
+    return dt_obj
+end
+
+local function datetime_new_dt(dt, secs, fraction, offset)
+    local epochV = dt ~= nil and (builtin.tnt_dt_rdn(dt) - DT_EPOCH_1970_OFFSET) *
```

Comment:
What is `V`? Also please, do not use camel-case. That comment definitely was already given in the previous version. Please, revisit the old comments.

Reply:
'V' here stands for value (got after nil handling). Got rid of V copies.